{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pysam\n",
      "  Downloading pysam-0.16.0.1-cp38-cp38-manylinux1_x86_64.whl (10.2 MB)\n",
      "\u001b[K     |████████████████████████████████| 10.2 MB 17.9 MB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: pysam\n",
      "Successfully installed pysam-0.16.0.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = pysam.AlignmentFile(\"/sbgenomics/project-files/merged-tumor.bam\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The first read from the file:\n",
      " C0HVYACXX120402:7:1207:5722:57044\t1187\t20\t9483248\t27\t76M\t20\t9483381\t76\tTTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG\tarray('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])\t[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]\n",
      "The flag field of the first read: 1187\n",
      "The number of unmapped reads: 17765\n",
      "The total number of reads: 2921629\n",
      "The number of reads with mapping quality 0: 126628\n",
      "The average mapping quality for all the reads: 55.91379158681681\n",
      "The average mapping quality if reads with 0 mapp quality are filtered out: 58.446975510921106\n"
     ]
    }
   ],
   "source": [
    "first = 0\n",
    "cnt_unmapped_reads = 0\n",
    "cnt_total_reads = 0\n",
    "cnt_mapping_quality_0 = 0\n",
    "mapping_quality_sum = 0\n",
    "\n",
    "for read in file:\n",
    "    if first == 0:\n",
    "        first_read = read\n",
    "        first = -1\n",
    "    if read.is_unmapped:\n",
    "        cnt_unmapped_reads += 1\n",
    "    cnt_total_reads += 1\n",
    "    if read.mapping_quality == 0:\n",
    "        cnt_mapping_quality_0 += 1\n",
    "    mapping_quality_sum += read.mapping_quality\n",
    "\n",
    "average_mapping_quality = mapping_quality_sum / cnt_total_reads\n",
    "average_mapping_quality_without_0 = mapping_quality_sum / (cnt_total_reads - cnt_mapping_quality_0)\n",
    "\n",
    "print(f\"The first read from the file:\\n {first_read}\")\n",
    "print(f\"The flag field of the first read: {first_read.flag}\")\n",
    "print(f\"The number of unmapped reads: {cnt_unmapped_reads}\")\n",
    "print(f\"The total number of reads: {cnt_total_reads}\")\n",
    "print(f\"The number of reads with mapping quality 0: {cnt_mapping_quality_0}\")\n",
    "print(f\"The average mapping quality for all the reads: {average_mapping_quality}\")\n",
    "print(f\"The average mapping quality if reads with 0 mapp quality are filtered out: {average_mapping_quality_without_0}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
